Pahole optimization for resolved's DnsQuery

keszybz opened the pull request.

bluca reviewed (approved):
LGTM, and the packit failure is unrelated (weird one, fails to generate man/systemd.directives.xml)

I wonder if it would be worth at some point to add automated checks for this kind of optimizations, for the structures in critical perf paths?

keszybz commented:
> LGTM, and the packit failure is unrelated (weird one, fails to generate man/systemd.directives.xml)

https://github.com/systemd/systemd/issues/19601

> I wonder if it would be worth at some point to add automated checks for this kind of optimizations, for the structures in critical perf paths?

I don't think so. It doesn't make sense to optimize singletons or structures which see very small use… but it'd be hard to automatize this check. So I think it's if we adjust it by hand every once in a while.

poettering commented on src/resolve/resolved-dns-query.h:
```diff
@@ -38,7 +38,7 @@ struct DnsQuery {
         DnsQuery *auxiliary_for;
         LIST_HEAD(DnsQuery, auxiliary_queries);
-        unsigned n_auxiliary_queries;
+        uint8_t n_auxiliary_queries;
```
what's the benefit of this? i.e. this is placed between a native length ptr and a 32bit int. thus, on 32bit this will begin on 32bit boundary and the next field starts at 32bit, so using 32bit should be fine. on 64bit this will begin on 64bit boundary and the next field starts 32bit later, so pretty much the same story.

poettering commented on src/resolve/resolved-dns-query.h:
```diff
@@ -64,7 +64,7 @@ struct DnsQuery {
 
         DnsTransactionState state;
-        unsigned n_cname_redirects;
+        uint8_t n_cname_redirects;
```
similar. this is between a 32bit enum, and a native width ptr. so using 32bit shuld be fine...

poettering commented:
ah, so the fourth commit explains the first two, due to the reordering. Any chance you can add a comment about this to the two earlier commits, so make it clear that it's worth shortening the types due to the follow-up commit that reorders stuff to fill gaps better

poettering commented on src/resolve/resolved-dns-query.h:
```diff
@@ -33,14 +33,6 @@ struct DnsQueryCandidate {
 struct DnsQuery {
```
maybe add a comment here somewhere that one is supposed to look at pahole before adding/rearranging the structs blindly

keszybz commented:
Updated. I added comments to the commits and to the structures, as requested. I also added a few more commits in the same vein. I'm unsetting the green labels because it probably needs another look.
